Cloudinit- terraform

MalloZup opened the pull request:
Ubuntu 18.04 with cloudinit

## TOdos: 
- [x] after discussion on irc would be nice to have a DISK resize to 200G for ubuntu image.


## Requirements of this PR:

- we need the `0.5.0` version of terraform-libvirt-provider in sumaform before merging this.

MalloZup commented:
@moio  with the actual Patch i was able to have a ubuntu with QEMU-AGENT and the saltstates were executed successfully.

Now the question we have to think are following about design/architecture of this integration:

1) cloudinit: one generic one file or one cloudinit-file pro DISTRO? i would prefer to have the generic one. ( but some pkgs like`avahi` change from DIstro to distro.

I think having 1 cloudinit for all would avoid maintainance and inputs bugs etc.

2) Actually at this point we generate a ISO for each VM we onboard. At this point we waste some small space because we generate ISO (cloudinit).

I want to keep this approach because we want to migrate all clients to CloudInit so this is the right approach (even if we are on transition).


3) https://github.com/moio/sumaform/pull/432/files#diff-299bef565123dc0baa9415daf859349fR64

moio commented:
You can probably construct the cloud init string directly in HCL, adding conditionals where strictly needed. I agree keeping the file as similar as possible across distros would be a huge win.

Can you add SLES JeOSs here?

It... Should... Work out of the box!

MalloZup commented:
i will try the JEOS. Afaik Leap15 as bug with Cloudinit, i spoke with bernard time ago.

So indeed JEOS is better :) After adding JEOS i would have the approach to merge this PR and not have a big Patch.

moio commented:
Not sure, we will see

MalloZup commented:
ok i will have better picture once i add the JEOS.
if you have any remark on actual codebase feel free to share. In the design architecture and machinery of terraform logic and also placement of file. I think it is ok but your feedback is always welcome :+1: :1st_place_medal: :heart_eyes: :guitar: :sunflower:

MalloZup commented:
```
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.controller_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.131922
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.327 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.default_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.133759
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.347 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.evil_minions_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.135548
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.366 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.grafana_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.137339
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.441 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: default_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.139234
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.331 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.minionswarm_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.140956
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.466 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.suse_manager_proxy_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.142819
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.316 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.suse_manager_server_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.144661
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.578 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos.testsuite_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.146718
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.592 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: repos_nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: test.nop
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Success!
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.148988
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 1.452 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: kernel_hostname
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: cmd.run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: sysctl kernel.hostname=dam-min-ubuntu
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: unless execution succeeded
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.151841
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 11.479 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: temporary_hostname
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: cmd.run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: hostnamectl set-hostname dam-min-ubuntu
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Command "hostnamectl set-hostname dam-min-ubuntu" run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.163973
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 12.428 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               pid:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   3042
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               retcode:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   0
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               stderr:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               stdout:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: permanent_hostname
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.managed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/hostname
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: File /etc/hostname is in the correct state
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.184067
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 15.453 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: legacy_permanent_hostname
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.managed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/HOSTNAME
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: File /etc/HOSTNAME is in the correct state
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.200022
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 2.933 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: hosts_file_hack
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: cmd.script
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: salt://default/set_ip_in_etc_hosts.py.jinja
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Command 'salt://default/set_ip_in_etc_hosts.py.jinja' run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:07.203483
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 31.39 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               pid:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   3043
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               retcode:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   126
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               stderr:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   /bin/bash: /tmp/__salt.tmp.f16fmuiq.jinja: /usr/bin/python: bad interpreter: No such file or directory
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               stdout:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: pkg.installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Problem encountered installing package(s). Additional info follows:

module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               errors:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   - Running scope as unit: run-r0bd7b5277c064a32be48f6def942f95e.scope
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                     E: Unable to locate package avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:11.880726
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 4249.018 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.replace
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/avahi/avahi-daemon.conf
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: No changes needed to be made
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.130368
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 8.745 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: service.running
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: avahi-daemon
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: One or more requisite failed: default.avahi.avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: avahi_resolution_configuration
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.replace
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/nsswitch.conf
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: One or more requisite failed: default.avahi.avahi
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: mdns_allow_file
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.append
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/mdns.allow
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: File /etc/mdns.allow is in correct state
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.143949
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 7.953 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: systemd_machine_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: cmd.run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: rm -f /etc/machine-id && rm -f /var/lib/dbus/machine-id && dbus-uuidgen --ensure && systemd-machine-id-setup && touch /etc/machine-id-already-setup
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: onlyif execution failed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.152471
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 10.21 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: dbus_machine_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: cmd.run
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: rm -f /var/lib/dbus/machine-id && dbus-uuidgen --ensure
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: /var/lib/dbus/machine-id exists
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.163461
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 9.615 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minion_id_cleared
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.absent
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/salt/minion_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Removed file /etc/salt/minion_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.173812
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 2.063 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               removed:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   /etc/salt/minion_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: timezone_package
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: pkg.installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: tzdata
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: All specified packages are already installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.176305
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 10.748 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: timezone_symlink
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.symlink
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/localtime
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Created new symlink /etc/localtime -> /usr/share/zoneinfo/Europe/Berlin
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.188661
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 3.316 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               new:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   /etc/localtime
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: timezone_setting
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: timezone.system
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: Europe/Berlin
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Timezone Europe/Berlin already set, UTC already set to Europe/Berlin
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.194305
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 62.986 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: authorized_keys
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.append
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /root/.ssh/authorized_keys
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Appended 2 lines
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.257963
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 39.416 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               diff:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   --- 
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   +++ 
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   @@ -0,0 +1,2 @@
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   +ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCyuJPNE3NxGd7nV1nLTTB2z2a98dbkLc0aTjW7Wc151g9NWxiFNlvcnXKOlryXUDGSoNwnkFJ3ji/6B0rViDnkT1QCtzIwZJa2ZrTc2xjUKlnQeQEAsMnTybNrbtRb0dz9VPGxk+Ygn1krz3vbAyHNnTFTsLo5eoKJ+9S3vJ34d8p7+0O2Cxu9ZE5r3AJzuN/V8zea7rHZzpohWTRElAAR8plYWYJYkhvfLgGwqKgSFzTeEsIoiHLjdSNL3GFS6pDFeq6Eq92QLglDRIa1eYpFEwPiP2ItoJXWhNeZKR82WnehTs5pRwOWtD8QH5T2dlAIUiHZsaAQKrGqv2JKsXOh dario@dario-local
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   +ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDOSLK0kJAulwO9Qwm/DVKLsGzv8Fh+ZcGe5x5d+VzO4/+xKnxjOPvGVngm9Hki8N436bCCvIRqwzstDUyJnTHhEe3qzsl02QGqIIfVjzK3BEp4sA8XcmByBy3pxqbTjAjrW8a+n43DBRMx4J86LgywX3fFt8ceEym3rFer/IzQD9QvKRocycNCwhpmGKEDh4zdl9PdyU5RcUpiA01AD/8rdJMNYnUxIsaNElcgP2LSzROjNdyu11cq9h3/cb3b7oRRef718GOkIkzqL/szo4Mm/j4KfyG0SBWyPxlJktFADD3lcQIRrGakFhiZOFbg2HZci26q+l/hUfLMbzcBrWlc+gzY4z4LMF8YBy1x/x+yR4j1SzZuqempSRu7rDKevcIGBF/fALPyiXH2EzYKsVcN9nv5mFCjmwVK7WTjw1SFzufKwdmnvOYgyeYuMaOJv3LEBY1jJg8r4xqzuQx32KNK/iJz1E+njptV3ZhOYDiHBmieykP08nWxdS79N1DbljxRzS54T7CArfGo23nYTNlMewPeKFx7ZqtUVv03g8i1Hn8WdArz2TYciWWrPErBEALBpXsu7Oq5AT97wAhcAIf3B3dV5p/bXgnhALA/SodNBf/wr0U2f5VzTDyJ9eNZURt16jSxDgXkuKKY/rimLqqt2FC1SeicmTql1nVyHG3j9Q== root@controller
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minion_cucumber_requisites
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: pkg.installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: Problem encountered installing package(s). Additional info follows:

module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               errors:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   - Running scope as unit: run-raea029ab7d794edda4ce8ee7dbc4dc5d.scope
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                     E: Unable to locate package openscap-utils
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:16.298161
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 863.008 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minion_package
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: pkg.installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: salt-minion
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: All specified packages are already installed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:17.162116
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 9.952 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minion_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: file.managed
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: /etc/salt/minion_id
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: File /etc/salt/minion_id updated
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:17.172547
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 4.609 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):               diff:
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):                   New file
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minion_service
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: service.running
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):         Name: salt-minion
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: True
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: The service salt-minion is already running
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:17.178570
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 43.89 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): ----------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):           ID: minimal_package_update
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Function: pkg.latest
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):       Result: False
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Comment: No information found for 'salt'.
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Started: 13:50:17.223088
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):     Duration: 559.322 ms
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec):      Changes:   
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): 
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): Summary for local
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): -------------
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): Succeeded: 26 (changed=6)
module.min-ubuntu.module.minion.libvirt_domain.domain (remote-exec): Failed:     6
```

MalloZup commented:
@moio jfyi: at moment i tried out the JEOS sles12sp3 image the states were ok.

However i think adding the JEOS images for all oses is a big PR with the ubuntu one. 
( i had today another issue).

Adding the JEOS images will also impy that we need to run the testsuite against at least for SLE15 and SLE12sp3/2

To me from logical/pragmatic perspective i am hesitating to add JEOS images to this PRs. 
I will continue today with the experimentation with JEOS and UBUNTU  and we also need to consider (CENTOS7)

I think adding UBUNTU is really important and doesn't harm to the current design. 
Adding JEOS and CENTOS7 is a big step ( we need also to add the testsuite/ci support tests before migrating).

I will continue today to see more or less, but i am still a fun for the 1 ubuntu PR ( also for my dayly work i could only focus on that and then after we could have a look at JEOS).

MalloZup commented:
@moio dunno if you listed the research mailing list but the JEOS SLES12sp2 was never released.

This literally break our design of having everything from upstream .. :crying_cat_face:  Idea, thoughts? 

We might keep the sles12sp2 as non JEOS images and builded with KIWI. I am not happy but there is no alternative afaik

moio commented:
Argh, that is not good. We could at least use cloud-init in Kiwi generated images, so that we rely on one tool.

On OSs that have JeOS or similar images, we use the upstream ones, when not available, we use our own images.

"It should be just a matter of adding the cloud-init package" :crossed_fingers:

MalloZup commented:
@moio feel free to take a review on this PR. i will not add the JEOS since it is to much work and risks.

I will adress the comments. thanks! :sun_with_face:

moio commented:
@MalloZup does this need a newer version of `terraform-provider-libvirt` than the one in 

https://download.opensuse.org/repositories/systemsmanagement:/sumaform/openSUSE_Leap_15.0/x86_64/

?

moio commented:
Travis CI fails, BTW

MalloZup commented:
Yop we need the latest release o.5 for the cloudinit feature. I reset this to wip for the consoles, I need to check for the sumaform images

MalloZup commented:
@moio  can you update the latest sumaform pkg with the released provider? this would help! thx.

the released version is here: ( there is no breaking change for sumaform users)

https://software.opensuse.org/download/package?project=systemsmanagement:terraform&package=terraform-provider-libvirt

MalloZup commented:
@moio i tested this locally and there is no side effect on sles12sp3 clients and either cento7. this works.

We can merge it when you update sumaform PKG. Feel free to review it or ping me for any info

MalloZup commented:
after discussion on irc would be nice to have a DISK resize to 200G for ubuntu image.

moio commented on modules/libvirt/base/main.tf:
```diff
@@ -65,6 +65,14 @@ resource "libvirt_volume" "sles-es7_volume" {
+# We fetch the latest ubuntu release image from their mirrors
+resource "libvirt_volume" "ubuntu-1804_volume" {
+  name = "${var.name_prefix}ubuntu-1804"
+  source = "https://cloud-images.ubuntu.com/releases/18.04/release/ubuntu-18.04-server-cloudimg-amd64.img"
```
Why `server` and not `minimal`?

http://cloud-images.ubuntu.com/minimal/releases/bionic/release/

MalloZup replied:
i can try that out

moio replied:
That would be awesome :+1:

moio commented on modules/libvirt/host/cloud_init.cfg:
```diff
@@ -0,0 +1,31 @@
+disable_root: false
+
+# NOTE: This is distro specific.
+apt_update: true
```
Is this silently ignored on platforms that do not support this flag?

MalloZup replied:
at moment i think that our SLES images are not using `cloudinit`, even if we pass it throug.

To me, i think that when we switch to JEOS/SLES we will need to have 2 different cloudinit files because of the pkg names differs.
Imho this is something that we can postpone and openining an ISSUE as `follow-up` of this

MalloZup replied:
not only the pkgs name but i think other things differ so we might have 2 files in case

moio commented on modules/libvirt/host/cloud_init.cfg:
```diff
@@ -0,0 +1,31 @@
+
+
+runcmd:
+ - sed  -i '/PermitRootLogin/s/.*/PermitRootLogin yes/' /etc/ssh/sshd_config 
```
Isn't this `ssh_pwauth` + `disable_root`?

MalloZup replied:
it doesn't work for SLES JEOS.. so we might remove it since is ubuntu specific ( i think should work)

i might also change the name to `cloud_init.ubuntu`

MalloZup commented on modules/libvirt/host/main.tf:
```diff
@@ -15,6 +15,16 @@ resource "libvirt_volume" "main_disk" {
+  template = "${file("${path.module}/cloud_init.cfg")}"
+}
+
+resource "libvirt_cloudinit_disk" "minimalconf" {
```
do you know the `hcl` way for that? or link for taking example of it?

moio replied:
Using count as usual? You did it yourself in this very PR!

https://github.com/moio/sumaform/pull/432/files#diff-51d7f10ca260c873b1b44239232e4722R72

:wink:

MalloZup replied:
@moio hcl is not intuitive and just keep forgeting it :smile:  hopefully in the 12 thing will litle more better due to some improvements..

Bischoff replied:
The base problem its it's a declarative language, thus leading to unnatural and easy-to-forget workarounds like `count`... 🙊

MalloZup commented on modules/libvirt/host/main.tf:
```diff
@@ -23,6 +35,7 @@ resource "libvirt_domain" "domain" {
   count = "${var.count}"
   qemu_agent = true
 
+  cloudinit = "${libvirt_cloudinit_disk.minimalconf.id}"
```
@moio  this works only for `ubuntu` if we add a sles client

```hcl

* module.minssh-sles12sp3.libvirt_domain.domain: Resource 'libvirt_cloudinit_disk.minimalconf' not found for variable 'libvirt_cloudinit_disk.minimalconf.id'


```

MalloZup replied:
this is due because we create it only for ubuntu and it can be that ubuntu hasn't create the disk..

moio replied:
In terraform 0.12 ternary conditions will not evaluate both sides so this becomes a possibility.

For now, I think this will only work if the provider accepts an array, like the `network_interface` or `disk` attributes.

MalloZup replied:
on the provider cloudinit is not `typelist`. https://github.com/dmacvicar/terraform-provider-libvirt/blob/c37707c37dd95b2e941888073192eeb69a24c78f/libvirt/resource_libvirt_domain.go#L98

To me the pragmatic way at this time is that we waste a little space ( creating cloudinit.disk) for each image and for terraform v12 we can change this. ( we open an issue as reminder).

Changing the libvirt-provider for such a feature is not worth and will take more time then expected.. :smile: 

I think comparing back in the time where we create downladed all qcow for all clients, we can tollorate this and user will even don\t remark that. 

Just some numbers here:

`-rw-r--r-- 1 qemu qemu 364K Oct 17 17:05 sles12sp3_jeos_cloudinit.iso`

Bischoff replied:
slice() or ternary... no if's, it's a declarative language 🙉

moio replied:
I think I will ultimately accept this temporarily (until terraform 0.12, or all images really use cloud-init, whichever comes first).

MalloZup commented:
I think having some kb more is not a pb @moio

MalloZup commented:
Ping

moio commented:
Plan is to have a look today

moio commented:
Travis is broken BTW, please fix it

MalloZup commented:
I cannot fix Travis.

Bischoff commented on main.tf.libvirt.example:
```diff
@@ -46,3 +46,15 @@ module "minion" {
   // see modules/libvirt/minion/variables.tf for possible values
 }
+
+module "min-ubuntu" {
```
for the other ones, we have the version embedded:
```
module "min-sles12sp3" {
module "min-centos7" {
```

do the same, for example `min-ubuntu1804` ?

Bischoff commented on modules/libvirt/base/main.tf:
```diff
@@ -11,7 +11,7 @@ resource "libvirt_volume" "centos7_volume" {
 resource "libvirt_volume" "opensuse423_volume" {
   name = "${var.name_prefix}opensuse423"
-  source = "https://download.opensuse.org/repositories/systemsmanagement:/sumaform:/images:/libvirt/images/opensuse423.x86_64.qcow2"
+  source = "http://download.opensuse.org/repositories/systemsmanagement:/sumaform:/images:/libvirt/images/opensuse423.x86_64.qcow2"
```
Why no `https` ?

If unrelated to ubuntu but needed, maybe in a different PR...

Bischoff commented on modules/libvirt/base/variables.tf:
```diff
@@ -72,6 +72,6 @@ variable "bridge" {
 variable "images" {
   description = "list of images to be uploaded to the libvirt host, leave default for all"
-  default = ["centos7", "opensuse423",  "sles15",  "sles11sp4",  "sles12", "sles12sp1", "sles12sp2", "sles12sp3", "sles-es7"]
+  default = ["centos7", "opensuse423", "ubuntu-1804",  "sles15",  "sles11sp4",  "sles12", "sles12sp1", "sles12sp2", "sles12sp3", "sles-es7"]
```
Alphabetical order: move it after sles ?

Bischoff commented on modules/libvirt/host/main.tf:
```diff
@@ -50,6 +63,29 @@ resource "libvirt_domain" "domain" {
+      type        = "pty"
+      target_type = "virtio"
+      target_port = "1"
+  }
```
that one in a separate PR if we get it to work... 😆

moio replied:
Just tested on an existing mirror (after the VM has been re-built with this PR code, without changes to the volumes): `virsh console` works as expected.

moio commented:
I am not totally happy with this PR, as we currently use different mechanisms for different OSs (custom images without cloud-init and upstream images with cloud-init in libvirt, custom images with cloud-init in openstack, custom images without cloud-init in AWS).

Nevertheless we need this and we have to start somewhere, so I am accepting.

@MalloZup please see the commits I added on top, which among other things address all comments from @Bischoff.

Thanks a lot for getting this off the ground! :rocket:
